src/metrics/include/metrics/v2/event.hpp

```diff
@@ -247,70 +285,269 @@ class Collector {
+struct MetricsArgs {
+    MetricsArgs(const typename ClockSource::time_point& finish,
+                OperationEventT<ClockSource> event,
+                size_t workerCount)
+        : finish(finish), event(std::move(event)), workerCount(workerCount) {}
+    typename ClockSource::time_point finish;
+    OperationEventT<ClockSource> event;
+    size_t workerCount;
+};
+
+// Efficient buffer that should rarely be blocked on insertions.
+template <typename ClockSource>
+class MetricsBuffer {
+public:
+    explicit MetricsBuffer(size_t size, const std::string& name)
+        : size{size},
+          name{name},
+          _loading(new std::vector<MetricsArgs<ClockSource>>()),
+          _draining(new std::vector<MetricsArgs<ClockSource>>()) {
+        _draining->reserve(size);
+        _loading->reserve(size);
+    }
+
+    // Thread-safe.
+    void addAt(const typename ClockSource::time_point& finish,
+               OperationEventT<ClockSource> event,
+               size_t workerCount) {
+        _loading_mutex.lock();
+        _loading->emplace_back(finish, std::move(event), workerCount);
+        _loading_mutex.unlock();
+    }
+
+    // Not thread-safe.
+    std::optional<MetricsArgs<ClockSource>> pop(bool force) {
+        refresh(force);
+        if (_draining->empty()) {
+            return std::nullopt;
+        } else {
+            auto ret = std::move(_draining->back());
+            _draining->pop_back();
+            return ret;
+        }
+    }
+
+    const std::string name;
+    const size_t size;
+
+private:
+    void refresh(bool force) {
+        if (_draining->empty()) {
+            _loading_mutex.lock();
+            if (force || _loading->size() >= size * SWAP_BUFFER_PERCENT) {
+                _draining.swap(_loading);
+            }
+            _loading_mutex.unlock();
+        }
+
+        // Maybe a bit nuclear, but this draws a box around the entire grpc system
+        // and errors if it ever backs up enough to slow down an actor thread.
+        if (_draining->size() > size) {
+            std::ostringstream os;
+            os << "Metrics buffer for operation name " << name << " exceeded pre-allocated space"
+               << ". Expected size: " << size << ". Actual size: " << _draining->size()
+               << ". This may affect recorded performance.";
+
+            BOOST_THROW_EXCEPTION(MetricsError(os.str()));
+        }
+    }
+
+    std::unique_ptr<std::vector<MetricsArgs<ClockSource>>> _loading;
+    std::unique_ptr<std::vector<MetricsArgs<ClockSource>>> _draining;
+    std::mutex _loading_mutex;
+};
 /**
  * Primary point of interaction between v2 poplar internals and the metrics system.
  */
 template <typename ClockSource, typename StreamInterface>
 class EventStream {
     using duration = typename ClockSource::duration;
     using OptionalPhaseNumber = std::optional<genny::PhaseNumber>;
+    using ClientPtr = std::shared_ptr<GrpcClient<ClockSource, StreamInterface>>;
```

Comment:
The use of `shared_ptr` is really the only thing giving me pause about giving my LGTM here. There are valid uses for it, but I don't quite see it. If it is indeed the right decision, please add a bit of commentary on the justification.